docs/dev/README.md

```diff
@@ -0,0 +1,218 @@
+```
+
+Your terminal looks something like this when the virtual environment is active:
+```
+(.venv) $
+```
+Now any time you type `python3` or `pip` or even `python`, you are using the
+one from your virtual environment.
+To stop using the virtual environment, run `deactivate`.
+
+## Install
+
+Ensure your tools are up to date:
+```sh
+(.venv) $ python3 -m pip install --upgrade pip setuptools
+```
+
+Install dev dependencies:
+```sh
+(.venv) $ python3 -m pip install --requirement requirements-dev.txt
+```
+
+Install aws-crt-python (helper script `./scripts/install-dev.py` does this):
+```sh
+(.venv) $ python3 -m pip install --verbose --editable .
+```
+
+You must re-run this command any time the C source code changes.
+But you don't need to re-run it if .py files change
+(thanks to the `--editable` aka "develop mode" flag)
+
+Note that this takes about twice as long on Mac, which compiles C for both `x86_64` and `arm64`.
+(TODO: in develop mode, for productivity's sake, only compile C for one architecture,
+and ignore resulting warnings from the linker)
+
+## Run Tests
+
+To run all tests:
+```sh
+(.venv) $ python3 -m unittest discover --failfast --verbose
+```
+
+`discover` automatically finds all tests to run
+
+`--failfast` stops after one failed test.
+This is useful because a failed test is likely to leak memory,
+which will cause the rest of the tests to fail as well.
+
+`--verbose` makes it easier to see which tests are skipped.
+
+To run specific tests, specify a path. For example:
+```sh
+(.venv) $ python3 -m unittest --failfast --verbose test.test_http_client.TestClient.test_connect_http
+```
+
+More path examples:
+*   test - everything under `test/` folder
+*   test.test_http_client - every test in `test_http_client.py`
+*   test.test_http_client.TestClient - every test in `TestClient` class
+*   test.test_http_client.TestClient.test_connect_http - A single test
+
+When creating new tests, note that the names of test files and test functions must be prefixed with `test_`.
+
+### Environment Variables for Tests
+Many tests require an AWS account. These tests are skipped unless
+specific environment variables are set:
+
+*   MQTT
+    *   AWS_TEST_IOT_MQTT_ENDPOINT - AWS account-specific endpoint to connect to IoT core by
+    *   AWS_TEST_TLS_CERT_PATH - file path to certificate used to initialize the TLS context of the MQTT connection
+    *   AWS_TEST_TLS_KEY_PATH - file path to the private key used to initialize the TLS context of the MQTT connection
+    *   AWS_TEST_TLS_ROOT_CERT_PATH - file path to the root CA used to initialize the TLS context of the MQTT connection
+*   PKCS11
+    *   AWS_TEST_PKCS11_LIB - path to PKCS#11 library
+    *   AWS_TEST_PKCS11_PIN - user PIN for logging into PKCS#11 token
+    *   AWS_TEST_PKCS11_TOKEN_LABEL - label of PKCS#11 token
+    *   AWS_TEST_PKCS11_KEY_LABEL - label of private key on PKCS#11 token,
+        which must correspond to the cert at AWS_TEST_TLS_CERT_PATH.
```

Comment:
Super minor nitpick: I'd remove the newline here and do it all on one line for consistency with the rest of the file.